Sendcoins

Chinwendu20 opened the pull request:
## Change Description
Description of change / link to associated issue.

## Steps to Test
Steps for reviewers to follow to test the change.

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

Chinwendu20 committed:
lnrpc: Add `Outpoint` field in `SendCoinsRequest`

Signed-off-by: Ononiwu Maureen <59079323+Chinwendu20@users.noreply.github.com>

Chinwendu20 committed:
itest: Update itest with sendcoins with utxo itest

Signed-off-by: Ononiwu Maureen <59079323+Chinwendu20@users.noreply.github.com>

hieblmi commented on cmd/lncli/commands.go:
```diff
@@ -311,16 +311,28 @@ var sendCoinsCommand = cli.Command{
+			Name: "utxo",
+			Usage: "a utxo specified as outpoint(tx:idx) which " +
+				"will be used to fund a channel. This flag " +
+				"can be repeatedly used to fund a channel " +
```
text needs adjustment once out of draft

hieblmi commented on cmd/lncli/commands.go:
```diff
@@ -393,6 +405,15 @@ func sendCoins(ctx *cli.Context) error {
 		displayAmt = balanceResponse.GetConfirmedBalance()
 	}
 
+	if ctx.IsSet("utxo") {
```
we need to make sure that there is no conflict with other flags here, like `sendall` and `amt`. We should fail early if there's a combination that doesn't make sense.

Chinwendu20 replied:
I think there is a check down there already.

hieblmi commented on cmd/lncli/commands.go:
```diff
@@ -393,6 +405,15 @@ func sendCoins(ctx *cli.Context) error {
+	if ctx.IsSet("utxo") {
+		utxos := ctx.StringSlice("utxo")
+
+		outpoints, err = utxosToOutpoints(utxos)
```
maybe we can move `utxosToOutpoints` to `types.go`.

hieblmi commented on itest/lnd_channel_funding_utxo_selection_test.go:
```diff
@@ -24,7 +24,7 @@ type chanFundUtxoSelectionTestCase struct {
 	selectedCoins []btcutil.Amount
 
-	// localAmt is the local portion of the channel funding amount.
+	// amt is the local portion of the channel funding amount.
```
revert

hieblmi commented on rpcserver.go:
```diff
@@ -1283,16 +1287,25 @@ func (r *rpcServer) SendCoins(ctx context.Context,
+	lnWallet := r.server.cc.Wallet
 	maxFeeRate := r.cfg.Sweeper.MaxFeeRate.FeePerKWeight()
 
+	var selectOutpoints []wire.OutPoint
```
should move this block to where it is used, down in the else branch. To make this easier to read we might want to refactor the if/else into two separate methods.

hieblmi commented on rpcserver.go:
```diff
@@ -1283,16 +1287,25 @@ func (r *rpcServer) SendCoins(ctx context.Context,
 	var txid *chainhash.Hash
 
-	wallet := r.server.cc.Wallet
+	lnWallet := r.server.cc.Wallet
```
I guess this can be reverted?

hieblmi commented on rpcserver.go:
```diff
@@ -1409,15 +1423,15 @@ func (r *rpcServer) SendCoins(ctx context.Context,
 		// selection (funding, sweep alls, other sends) can proceed
-		// while we instruct the wallet to send this transaction.
+		// while we instruct the lnWallet to send this transaction.
 		paymentMap := map[string]int64{targetAddr.String(): in.Amount}
```
the `in.Amount` here needs to point to the sum of values of all selected utxos

Chinwendu20 replied:
why would that be?

hieblmi commented on lnwallet/btcwallet/btcwallet.go:
```diff
@@ -997,7 +997,7 @@ func (b *BtcWallet) SendOutputs(outputs []*wire.TxOut,
 	return b.wallet.SendOutputs(
 		outputs, nil, defaultAccount, minConfs, feeSatPerKB,
-		b.cfg.CoinSelectionStrategy, label,
+		b.cfg.CoinSelectionStrategy, label, selectedUtxos...,
```
here we need to pass in a new strategy if utxos are selected, something that indicates manual selection

Chinwendu20 replied:
There is a new PR on lnd to add strategy as an arg to functions like this but for now I just skipped calling `ArrangeCoins` if there is a selected utxo in the btcd side.

hieblmi commented on lnwallet/btcwallet/btcwallet.go:
```diff
@@ -979,7 +979,7 @@ func (b *BtcWallet) ImportTaprootScript(scope waddrmgr.KeyScope,
 func (b *BtcWallet) SendOutputs(outputs []*wire.TxOut,
 	feeRate chainfee.SatPerKWeight, minConfs int32,
-	label string) (*wire.MsgTx, error) {
+	label string, selectedUtxos ...[]wire.OutPoint) (*wire.MsgTx, error) {
```
This should be of type `...wire.OutPoint`

hieblmi reviewed (commented):
Very nice, some comments I have so far.

Chinwendu20 committed:
lntest+lnwallet: Adapt walletctrl, UTXO sendcoins

Signed-off-by: Ononiwu Maureen <59079323+Chinwendu20@users.noreply.github.com>

Chinwendu20 committed:
cmd: Export and move `utxosToOutpoint`.

Signed-off-by: Ononiwu Maureen <59079323+Chinwendu20@users.noreply.github.com>

Chinwendu20 commented:
I really appreciate your review @hieblmi, I have updated lnd with the PR, we can continue convo there.

Chinwendu20 committed:
sweep: Add selectUtxos to CraftSweepAllTx args

Signed-off-by: Ononiwu Maureen <59079323+Chinwendu20@users.noreply.github.com>

Chinwendu20 committed:
lnd: modify sendcoin rpc impl for select utxos

Signed-off-by: Ononiwu Maureen <59079323+Chinwendu20@users.noreply.github.com>

Chinwendu20 committed:
lncli: Add `outpoints` flag to sendcoins command

Signed-off-by: Ononiwu Maureen <59079323+Chinwendu20@users.noreply.github.com>

ProofOfKeags committed:
itest+lntest: add itest to reproduce bug #8535

ProofOfKeags committed:
contractcourt: consider delivery addresses when evaluating toSelfAmount

This commit fixes #8535 by changing how we assess toSelfAmount inside
the chainWatcher.

In certain cases users may wish to close out channel funds to external
delivery addresses set either during open or close.

Prior to this change we only consider addresses that our wallet is
aware of.

This change now identifies outputs as to_self outputs if the delivery
script matches OR if our wallet is aware of the address. In certain
edge cases it can be possible for there to be more than one output
that matches these criteria and in that case we will return the sum
of those values.

Chinwendu20 committed:
itest+lntest: Change `SendCoinsAssertErr`

This commit changes `SendCoinsAssertErr` to `SendCoinsReturnErr`. The
function now returns the resulting error from calling `sendCoins`.

Signed-off-by: Ononiwu Maureen <59079323+Chinwendu20@users.noreply.github.com>

ProofOfKeags committed:
docs: update release notes

Chinwendu20 committed:
docs: Added release note

Signed-off-by: Ononiwu Maureen <59079323+Chinwendu20@users.noreply.github.com>

yyforyongyu committed:
Merge pull request #8800 from ProofOfKeags/bugfix/8535

contractcourt: consider delivery addresses when evaluating toSelfAmount

Chinwendu20 committed:
gomod: fix lint

Signed-off-by: Ononiwu Maureen <59079323+Chinwendu20@users.noreply.github.com>
